src/main/java/com/faforever/api/avatar/AvatarService.java

```diff
@@ -0,0 +1,88 @@
+package com.faforever.api.avatar;
+
+import com.faforever.api.config.FafApiProperties;
+import com.faforever.api.data.domain.Avatar;
+import com.faforever.api.error.ApiException;
+import com.faforever.api.error.Error;
+import com.faforever.api.error.ErrorCode;
+import com.faforever.api.utils.FilePermissionUtil;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.dao.DataAccessException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+
+@Service
+@Slf4j
+public class AvatarService {
+  private final AvatarRepository avatarRepository;
+  private final FafApiProperties properties;
+
+  public AvatarService(AvatarRepository avatarRepository, FafApiProperties properties) {
+    this.avatarRepository = avatarRepository;
+    this.properties = properties;
+  }
+
+  @SneakyThrows
+  @Transactional
+  public void processUploadedAvatar(String tooltip, String originalFilename, InputStream imageDataInputStream, long avatarImageFileSize) {
+    String url = properties.getAvatar().getDownloadUrlBase() + originalFilename;
+    validateImageFile(originalFilename, avatarImageFileSize);
+
+    avatarRepository.findByUrl(url).ifPresent(avatar -> {
+      throw new ApiException(new Error(ErrorCode.AVATAR_NAME_CONFLICT, originalFilename));
+    });
+    final Path imageTargetPath = properties.getAvatar().getTargetDirectory().resolve(originalFilename);
+    Files.copy(imageDataInputStream, imageTargetPath, StandardCopyOption.REPLACE_EXISTING);
+    FilePermissionUtil.setDefaultFilePermission(imageTargetPath);
+    final Avatar avatar = new Avatar()
+      .setTooltip(tooltip)
+      .setUrl(url);
+    try {
+      avatarRepository.save(avatar);
+    } catch (DataAccessException e) {
+      try {
+        Files.delete(imageTargetPath);
+      } catch (IOException ioException) {
+        log.warn("Could not delete file " + imageTargetPath, ioException);
+      }
+      throw e;
+    }
+  }
+
+  @SneakyThrows
+  @Transactional
+  public void deleteAvatar(Integer avatarId) {
+    final Avatar avatar = avatarRepository.getOne(avatarId);
```

Comment:
so I should not focus on returning 404? Or should I "hack" exception handler that in case of this particular ApiException the return code should be 404?

Reply:
Yes you should. But actually you are not checking the avatar == null. And since you already declared `findById` and it returns an optional, it's easier to use the optional.

Reply:
getOne will not return null. It will throw EntityNotFoundException which I handled as 404 in globalexceptionhandler and could be used by any other repository. In previous implementation I used findById but throw it away since throwing ApiException would result in 422 or I would need to hack handler method.
So what is the correct path I should choose?